### PR TITLE
Added the ability to make additional parameters when creating and upd…

### PR DESCRIPTION
**Pull request by MikhailKlimenko**
Greetings, recently I needed to develop a plugin for redmine and in the plugin I need to add additional parameters when updating and creating issue.

### CODE REVIEW

File: src/main/java/com/taskadapter/redmineapi/bean/Issue.java [hunk 1]
```diff
@@ -616,13 +613,15 @@ public PropertyStorage getStorage() {
...
      * @throws RedmineException
      */
-    public Issue create() throws RedmineException {
-        return transport.addObject(this, new BasicNameValuePair("include",
-                Include.attachments.toString()));
+    public Issue create(NameValuePair... params) throws RedmineException {
+        NameValuePair[] enrichParams = Arrays.copyOf(params, params.length + 1);
+        enrichParams[params.length] = new BasicNameValuePair("include",
+                Include.attachments.toString());
+        return transport.addObject(this, enrichParams);
```
Comment (alexeyOnGitHub) [inline src/main/java/com/taskadapter/redmineapi/bean/Issue.java]: 1. 
I am fine with the general idea, but let's remove apachehttp-specific NameValuePair from Issue class params.
I see so far NameValuePair is only used deep inside Transport internals, like URICOnfigurator or Transport class itself. 
it is better to keep "user-facing" classes like `Issue` free of transport details. we may want to replace Apache HttpCore library later.

2. 
need tests to ensure the methods work fine with no params, null param, *some* params, etc.

**Comment by MikhailKlimenko** [comment]
Hi, I've separated the transport details "NameValuePair" from user-facing, added a Converter and hashset to UriConfigurator to weed out null and similar keys.

**Comment by alexeyOnGitHub** [comment]
I am currently traveling, will take a look at this in about a week. Thank you for helping out with this library, Mikhail!

File: src/main/java/com/taskadapter/redmineapi/IssueManager.java [hunk 2]
```diff
@@ -9,10 +9,7 @@
 import com.taskadapter.redmineapi.bean.SavedQuery;
 import com.taskadapter.redmineapi.bean.Tracker;
 import com.taskadapter.redmineapi.bean.Watcher;
-import com.taskadapter.redmineapi.internal.DirectObjectsSearcher;
-import com.taskadapter.redmineapi.internal.Joiner;
-import com.taskadapter.redmineapi.internal.ResultsWrapper;
-import com.taskadapter.redmineapi.internal.Transport;
+import com.taskadapter.redmineapi.internal.*;
 import org.apache.http.NameValuePair;
```
Comment (alexeyOnGitHub) [inline src/main/java/com/taskadapter/redmineapi/IssueManager.java]: let's avoid using wildcard imports here as well as in other places, those bring random classes at times.

File: src/main/java/com/taskadapter/redmineapi/internal/RequestParam.java [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+        return value;
+    }
+
+    @Override
+    public boolean equals(Object o) {
+        if (this == o) return true;
+        if (o == null || getClass() != o.getClass()) return false;
+        RequestParam that = (RequestParam) o;
+        return name.equals(that.name);
+    }
```
Comment (alexeyOnGitHub) [inline src/main/java/com/taskadapter/redmineapi/internal/RequestParam.java]: this may be confusing to people. why are params with the same name considered "equal"? what about the value?

if some part of the code wants to treat these Params as "equal" as long as they have equal names, it is better to keep this comparison logic close to that area, maybe with a local Comparator class or something, but not inside the data object itself. this would allow reusability for this class in the future.

File: src/main/java/com/taskadapter/redmineapi/bean/Issue.java [hunk 4]
```diff
@@ -5,14 +5,12 @@
 import com.taskadapter.redmineapi.NotFoundException;
 import com.taskadapter.redmineapi.RedmineAuthenticationException;
 import com.taskadapter.redmineapi.RedmineException;
+import com.taskadapter.redmineapi.internal.RequestParam;
 import com.taskadapter.redmineapi.internal.Transport;
+import org.apache.http.NameValuePair;
 import org.apache.http.message.BasicNameValuePair;
```
Comment (alexeyOnGitHub) [inline src/main/java/com/taskadapter/redmineapi/bean/Issue.java]: these have become unused now

File: src/main/java/com/taskadapter/redmineapi/internal/URIConfigurator.java [hunk 5]
```diff
@@ -96,15 +92,16 @@ public URI createURI(String query, NameValuePair... param) {
...
-                origParams);
-        if (apiAccessKey != null) {
-            params.add(new BasicNameValuePair("key", apiAccessKey));
-        }
+                          Collection<? extends RequestParam> origParams) {
+
+//        final List<NameValuePair> params = new ArrayList<>(origParams);
+//        if (apiAccessKey != null) {
+//            params.add(new BasicNameValuePair("key", apiAccessKey));
+//        }
```
Comment (alexeyOnGitHub) [inline src/main/java/com/taskadapter/redmineapi/internal/URIConfigurator.java]: commented out code

File: src/main/java/com/taskadapter/redmineapi/internal/URIConfigurator.java [hunk 6]
```diff
@@ -118,41 +115,49 @@ private URI createURI(String query,
         }
     }
 
+    public Collection<? extends NameValuePair> convert(Collection<? extends RequestParam> origParams) {
```
Comment (alexeyOnGitHub) [inline src/main/java/com/taskadapter/redmineapi/internal/URIConfigurator.java]: this should be a private method. also, "convert" sounds very generic. something like "toNameValue" would be easier to read for people.
plus it should be static

File: src/main/java/com/taskadapter/redmineapi/internal/URIConfigurator.java [hunk 7]
```diff
@@ -96,15 +92,16 @@ public URI createURI(String query, NameValuePair... param) {
...
-        if (apiAccessKey != null) {
-            params.add(new BasicNameValuePair("key", apiAccessKey));
-        }
+                          Collection<? extends RequestParam> origParams) {
+
+//        final List<NameValuePair> params = new ArrayList<>(origParams);
+//        if (apiAccessKey != null) {
+//            params.add(new BasicNameValuePair("key", apiAccessKey));
+//        }
+        Set<RequestParam> requestParamSet = new HashSet<RequestParam>(origParams);
```
Comment (alexeyOnGitHub) [inline src/main/java/com/taskadapter/redmineapi/internal/URIConfigurator.java]: if this set creation is the main reason why RequestParam instances are  considered equal if their names are equal, it is better to keep the corresponding logic right here

File: src/test/java/com/taskadapter/redmineapi/IssueManagerIT.java [hunk 8]
```diff
@@ -303,6 +304,74 @@ public void testCreateIssueSummaryOnly() throws RedmineException {
...
+
+    @Test
+    public void testCreateIssueWithNullParams3() throws RedmineException {
+        RequestParam param1 = new RequestParam("name1", "param1");
+        RequestParam param3 = new RequestParam("name3", "param3");
+        Issue issue = new Issue(transport, projectId).setSubject("This is the Issue with null params")
+                .create(param1, null, param3);
+        assertNotNull("Checking returned result", issue);
+        assertNotNull("New issue must have some ID", issue.getId());
+    }
```
Comment (alexeyOnGitHub) [inline src/test/java/com/taskadapter/redmineapi/IssueManagerIT.java]: these positive tests have a lot of duplication in them.  it makes sense to extract the common parts into utility method(s)

File: src/main/java/com/taskadapter/redmineapi/internal/URIConfigurator.java [hunk 9]
```diff
@@ -118,41 +116,57 @@ private URI createURI(String query,
...
+    }
+
+    private static Collection<? extends NameValuePair> toNameValue(Collection<? extends RequestParam> origParams) {
+        return origParams
+                .stream()
+                .filter(Objects::nonNull)
+                .map(p -> new BasicNameValuePair(p.getName(), p.getValue()))
+                .collect(Collectors.toSet());
+    }
+
```
Comment (alexeyOnGitHub) [inline src/main/java/com/taskadapter/redmineapi/internal/URIConfigurator.java]: awesome, let's add basic unit tests for these two methods

**Comment by alexeyOnGitHub** [comment]
I made the final changes in your branch and merged it here : https://github.com/taskadapter/redmine-java-api/pull/335
thank you for helping out with this library!
